File: app/src/main/java/com/icebem/akt/activity/AboutActivity.java
package com.icebem.akt.activity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.snackbar.Snackbar;
import com.icebem.akt.BuildConfig;
import com.icebem.akt.R;
import com.icebem.akt.app.PreferenceManager;
import com.icebem.akt.util.AppUtil;
import com.icebem.akt.util.DataUtil;
import com.icebem.akt.util.IOUtil;

import org.json.JSONObject;

import java.io.IOException;

public class AboutActivity extends AppCompatActivity {
    private static final int TEXT_SPEED = 100;
    private int i;
    private TextView typeDesc, thanksDesc;
    private LinearLayout versionContainer;
    private PreferenceManager manager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_about);
        setSupportActionBar(findViewById(R.id.toolbar));
        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        findViewById(R.id.fab).setOnClickListener(this::onClick);
        initView();
    }

    private void initView() {
        findViewById(R.id.container_comment).setOnClickListener(this::onClick);
        findViewById(R.id.container_discuss).setOnClickListener(this::onClick);
        findViewById(R.id.container_project).setOnClickListener(this::onClick);
        versionContainer = findViewById(R.id.container_version_state);
        versionContainer.setOnClickListener(this::onClick);
        versionContainer.setOnLongClickListener(this::onLongClick);
        findViewById(R.id.container_version_type).setOnClickListener(this::onClick);
        findViewById(R.id.container_special_thanks).setOnClickListener(this::onClick);
        findViewById(R.id.green_android).setOnClickListener(this::onClick);
        ((TextView) findViewById(R.id.txt_version_state_desc)).setText(BuildConfig.VERSION_NAME);
        typeDesc = findViewById(R.id.txt_version_type_desc);
        thanksDesc = findViewById(R.id.special_thanks_desc);
        manager = PreferenceManager.getInstance(this);
        typeDesc.setText(manager.isPro() ? R.string.version_type_pro : R.string.version_type_lite);
    }

    private void onClick(View view) {
        switch (view.getId()) {
            case R.id.fab:
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(R.string.action_donate);
                builder.setMessage(R.string.msg_donate);
                builder.setNeutralButton(R.string.action_donate_payment, (dialog, which) -> {
                    Intent intent = null;
                    try {
                        intent = Intent.parseUri(AppUtil.URL_ALIPAY_API, Intent.URI_INTENT_SCHEME);
                    } catch (Exception e) {
                        Log.e(getClass().getSimpleName(), Log.getStackTraceString(e));
                    }
                    if (intent == null || intent.resolveActivity(getPackageManager()) == null)
                        intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_PAYPAL));
                    startActivity(intent);
                    Snackbar.make(view, R.string.info_donate_thanks, Snackbar.LENGTH_INDEFINITE).show();
                });
                builder.setPositiveButton(R.string.not_now, null);
                builder.setNegativeButton(R.string.no_thanks, null);
                builder.create().show();
                break;
            case R.id.container_comment:
                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_MARKET)).setPackage(AppUtil.MARKET_COOLAPK);
//                if (intent.resolveActivity(getPackageManager()) == null)
//                    intent.setPackage(AppUtil.MARKET_PLAY);
                if (intent.resolveActivity(getPackageManager()) == null)
                    intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_COOLAPK));
                startActivity(intent);
                break;
            case R.id.container_discuss:
                intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_QQ_API));
                if (intent.resolveActivity(getPackageManager()) == null)
                    intent.setData(Uri.parse(AppUtil.URL_PROJECT));
                startActivity(intent);
                break;
            case R.id.container_project:
                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_PROJECT)));
                break;
            case R.id.container_version_state:
                view.setClickable(false);
                view.setLongClickable(false);
                new Thread(this::checkVersionUpdate, AppUtil.THREAD_UPDATE).start();
                Snackbar.make(view, R.string.version_checking, Snackbar.LENGTH_INDEFINITE).show();
                break;
            case R.id.container_version_type:
                if (++i >= 15) {
                    i = 0;
                    manager.setPro(!manager.isPro());
                    typeDesc.setText(manager.isPro() ? R.string.version_type_pro : R.string.version_type_lite);
                    Snackbar.make(view, R.string.version_type_changed, Snackbar.LENGTH_LONG).show();
                }
                break;
            case R.id.container_special_thanks:
                if (++i >= 5) {
                    i = 0;
                    view.setClickable(false);
                    String extra = thanksDesc.getText() + System.lineSeparator() + System.lineSeparator() + getString(R.string.special_thanks_extra);
                    new Thread(() -> {
                        while (thanksDesc.getText().length() < extra.length()) {
                            thanksDesc.post(() -> thanksDesc.setText(extra.substring(0, thanksDesc.getText().length() + 1)));
                            SystemClock.sleep(TEXT_SPEED);
                        }
                    }, AppUtil.THREAD_UPDATE).start();
                }
                break;
            case R.id.green_android:
                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_GREEN_ANDROID)));
                break;
        }
    }

    private boolean onLongClick(View view) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.action_reset);
        builder.setMessage(R.string.msg_data_reset);
        builder.setPositiveButton(R.string.action_reset, (dialog, which) -> {
            int id = R.string.data_reset_done;
            try {
                DataUtil.updateData(manager, false);
            } catch (Exception e) {
                id = R.string.error_occurred;
            }
            Snackbar.make(view, id, Snackbar.LENGTH_LONG).show();
        });
        builder.setNegativeButton(android.R.string.cancel, null);
        builder.create().show();
        return true;
    }

    private void checkVersionUpdate() {
        int id = R.string.version_update;
        String l = null;
        String url = AppUtil.URL_RELEASE_LATEST;
        try {
            if (AppUtil.isLatestVersion()) {
                id = DataUtil.updateData(manager, true) ? R.string.data_updated : R.string.version_latest;
            } else {
                JSONObject json = new JSONObject(IOUtil.stream2String(IOUtil.fromWeb(AppUtil.URL_RELEASE_LATEST_API)));
                l = AppUtil.getChangelog(json);
                url = AppUtil.getDownloadUrl(json);
            }
            manager.setCheckLastTime();
        } catch (Exception e) {
            id = R.string.version_checking_failed;
            if (e instanceof IOException)
                l = getString(R.string.msg_network_error);
        }
        int result = id;
        String log = l;
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        typeDesc.post(() -> {
            if (result == R.string.version_update) {
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(result);
                builder.setMessage(log);
                builder.setPositiveButton(R.string.action_update, (dialog, which) -> startActivity(intent));
                builder.setNegativeButton(R.string.no_thanks, null);
                builder.create().show();
            } else if (log != null) {
                Snackbar.make(typeDesc, result, Snackbar.LENGTH_INDEFINITE).setAction(R.string.action_details, v -> AppUtil.showLogDialog(this, log)).show();
            } else Snackbar.make(typeDesc, result, Snackbar.LENGTH_LONG).show();
            versionContainer.setOnClickListener(v -> startActivity(intent));
        });
    }
}

File: app/src/main/java/com/icebem/akt/activity/AboutActivity.kt
package com.icebem.akt.activity

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.SystemClock
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.icebem.akt.BuildConfig
import com.icebem.akt.R
import com.icebem.akt.app.PreferenceManager
import com.icebem.akt.util.AppUtil
import com.icebem.akt.util.DataUtil
import com.icebem.akt.util.IOUtil
import org.json.JSONObject
import java.io.IOException

class AboutActivity : AppCompatActivity() {
    companion object {
        private const val TEXT_SPEED = 50L
    }

    private var i = 0
    private lateinit var typeDesc: TextView
    private lateinit var thanksDesc: TextView
    private lateinit var versionContainer: View
    private lateinit var manager: PreferenceManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_about)
        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        initView()
    }

    private fun initView() {
        findViewById<View>(R.id.fab).setOnClickListener { onClick(it) }
        findViewById<View>(R.id.container_comment).setOnClickListener { onClick(it) }
        findViewById<View>(R.id.container_discuss).setOnClickListener { onClick(it) }
        findViewById<View>(R.id.container_project).setOnClickListener { onClick(it) }
        versionContainer = findViewById(R.id.container_version_state)
        versionContainer.setOnClickListener { onClick(it) }
        versionContainer.setOnLongClickListener { onLongClick() }
        findViewById<View>(R.id.container_version_type).setOnClickListener { onClick(it) }
        findViewById<View>(R.id.container_special_thanks).setOnClickListener { onClick(it) }
        findViewById<View>(R.id.free_android).setOnClickListener { onClick(it) }
        findViewById<TextView>(R.id.txt_version_state_desc).text = BuildConfig.VERSION_NAME
        typeDesc = findViewById(R.id.txt_version_type_desc)
        thanksDesc = findViewById(R.id.special_thanks_desc)
        manager = PreferenceManager.getInstance(this)
        typeDesc.setText(if (manager.isPro) R.string.version_type_pro else R.string.version_type_lite)
    }

    private fun onClick(view: View) {
        when (view.id) {
            R.id.fab -> {
                AlertDialog.Builder(this).apply {
                    setTitle(R.string.action_support)
                    setMessage(R.string.msg_donate)
                    setPositiveButton(R.string.action_donate) { _, _ -> onDonate() }
                    setNeutralButton(R.string.action_share) { _, _ ->
                        (getSystemService(CLIPBOARD_SERVICE) as ClipboardManager).setPrimaryClip(ClipData.newPlainText(getString(R.string.app_name), AppUtil.URL_COOLAPK))
                        Snackbar.make(view, R.string.info_share, Snackbar.LENGTH_LONG).show()
                    }
                    setNegativeButton(R.string.not_now, null)
                    create().show()
                }
            }
            R.id.container_comment -> {
                var intent = Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_MARKET)).setPackage(AppUtil.MARKET_COOLAPK)
                if (intent.resolveActivity(packageManager) == null) intent = Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_COOLAPK))
                startActivity(intent)
            }
            R.id.container_discuss -> {
                intent = Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_QQ_API))
                if (intent.resolveActivity(packageManager) == null) intent.data = Uri.parse(AppUtil.URL_PROJECT)
                startActivity(intent)
            }
            R.id.container_project -> startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_PROJECT)))
            R.id.container_version_state -> {
                view.isClickable = false
                view.isLongClickable = false
                Thread({ checkVersionUpdate() }, AppUtil.THREAD_UPDATE).start()
                Snackbar.make(view, R.string.version_checking, Snackbar.LENGTH_INDEFINITE).show()
            }
            R.id.container_version_type -> if (++i >= 15) {
                i = 0
                manager.isPro = !manager.isPro
                typeDesc.setText(if (manager.isPro) R.string.version_type_pro else R.string.version_type_lite)
                Snackbar.make(view, R.string.version_type_changed, Snackbar.LENGTH_LONG).show()
            }
            R.id.container_special_thanks -> if (++i >= 5) {
                i = 0
                view.isClickable = false
                val extra = thanksDesc.text.toString() + System.lineSeparator() + System.lineSeparator() + getString(R.string.special_thanks_extra)
                Thread({
                    while (thanksDesc.text.length < extra.length) {
                        thanksDesc.post { thanksDesc.text = extra.substring(0, thanksDesc.text.length + 1) }
                        SystemClock.sleep(TEXT_SPEED)
                    }
                }, AppUtil.THREAD_UPDATE).start()
            }
            R.id.free_android -> startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_FREE_ANDROID)))
        }
    }

    private fun onLongClick(): Boolean {
        AlertDialog.Builder(this).apply {
            setTitle(R.string.action_reset)
            setMessage(R.string.msg_data_reset)
            setPositiveButton(R.string.action_reset) { _, _ ->
                var id = R.string.data_reset_done
                try {
                    DataUtil.updateData(context, null)
                    manager.setCheckLastTime(true)
                } catch (e: Exception) {
                    id = R.string.error_occurred
                }
                Snackbar.make(versionContainer, id, Snackbar.LENGTH_LONG).show()
            }
            setNegativeButton(android.R.string.cancel, null)
            create().show()
        }
        return true
    }

    private fun onDonate() {
        AlertDialog.Builder(this).apply {
            setTitle(R.string.action_donate)
            setSingleChoiceItems(resources.getStringArray(R.array.donate_payment_entries), 0) { dialog, which ->
                dialog.cancel()
                when (which) {
                    0 -> try {
                        startActivity(Intent.parseUri(AppUtil.URL_ALIPAY_API, Intent.URI_INTENT_SCHEME))
                    } catch (e: Exception) {
                        showQRDialog(true)
                    }
                    1 -> showQRDialog(false)
                    2 -> startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(AppUtil.URL_PAYPAL)))
                }
                Snackbar.make(typeDesc, R.string.info_donate_thanks, Snackbar.LENGTH_INDEFINITE).show()
            }
            setNegativeButton(R.string.no_thanks, null)
            create().show()
        }
    }

    private fun showQRDialog(isAlipay: Boolean) {
        AlertDialog.Builder(this).apply {
            setTitle(R.string.action_donate)
            setView(if (isAlipay) R.layout.qr_alipay else R.layout.qr_wechat)
            setPositiveButton(R.string.got_it, null)
            create().show()
        }
    }

    private fun checkVersionUpdate() {
        var id = R.string.version_update
        var log: String? = null
        var url = AppUtil.URL_RELEASE_LATEST
        try {
            val entry = DataUtil.getOnlineEntry()
            if (DataUtil.latestApp(entry)) {
                id = if (DataUtil.updateData(this, entry)) R.string.data_updated else R.string.version_latest
            } else {
                val json = JSONObject(IOUtil.stream2String(IOUtil.fromWeb(AppUtil.URL_RELEASE_LATEST_API)))
                log = DataUtil.getChangelog(json)
                url = DataUtil.getDownloadUrl(json)
            }
            manager.setCheckLastTime(false)
        } catch (e: Exception) {
            id = R.string.version_checking_failed
            if (e is IOException) log = getString(R.string.msg_network_error)
        }
        typeDesc.post {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            when {
                id == R.string.version_update -> {
                    AlertDialog.Builder(this).apply {
                        setTitle(id)
                        setMessage(log)
                        setPositiveButton(R.string.action_update) { _, _ -> startActivity(intent) }
                        setNegativeButton(R.string.no_thanks, null)
                        create().show()
                    }
                    Snackbar.make(typeDesc, id, Snackbar.LENGTH_LONG).show()
                }
                log != null -> Snackbar.make(typeDesc, id, Snackbar.LENGTH_INDEFINITE).setAction(R.string.action_details) { AppUtil.showLogDialog(this, log) }.show()
                else -> Snackbar.make(typeDesc, id, Snackbar.LENGTH_LONG).show()
            }
            versionContainer.setOnClickListener { startActivity(intent) }
        }
    }
}